src/main/src/main/java/org/geoserver/catalog/LayerGroupHelper.java

```diff
@@ -284,21 +321,82 @@ public String getLoopAsString(Stack<LayerGroupInfo> path) {
     private static boolean checkLoops(LayerGroupInfo group, Stack<LayerGroupInfo> path) {
         path.push(group);
         if (group.getLayers() != null) {
-            for (PublishedInfo child : group.getLayers()) {
+            int size = group.getLayers().size();
+            for (int i = 0; i < size; i++) {
+                PublishedInfo child = group.getLayers().get(i);
+                StyleInfo s;
+                //Handle incomplete layer groups, especially those constructed by the XStreamPersister
+                if (group.getStyles() == null || group.getStyles().size() == 0) {
+                    s = null;
+                } else {
+                    s = group.getStyles().get(i);
+                }
                 if (child instanceof LayerGroupInfo) {
                     if (isGroupInStack((LayerGroupInfo) child, path)) {
                         path.push((LayerGroupInfo) child);
                         return true;
                     } else if (checkLoops((LayerGroupInfo) child, path)) {
                         return true;
-                    }                
+                    }
+                } else if (child == null && s != null) {
+                    if (checkStyleGroupLoops(s, group, path)) {
+                        return true;
+                    }
                 }
             }
         }
         
         path.pop();
         return false;
     }
+
+    private static boolean checkStyleGroupLoops(StyleInfo styleGroup, LayerGroupInfo group, Stack<LayerGroupInfo> path) {
```

Comment:
looks like you are checking for recursion, can we have a java doc just to say what true and false mean?

Reply:
Sure. In that case I'll update `checkLoop` as well (the preexisting version of this)